emails: Inline CSS during the email building process.

shameondev opened the pull request:
This pull request is related to issue #24761.

Here's a quick overview of how it works after the changes:
1. The `inline_email_css.py` is no longer necessary for main email templates, so I have removed it.
1. Instead of using `inline_email_css`, we are now using the source versions of emails (which are not inlined) and inlining them directly during the email building process in `send_email.py` using the `get_inlined_template` function.
1. Added the `remove_style` argument to the `_test_cases` function, which allows for the removal of style attributes from the strings being tested. Additionally, the `remove_style_attribute` function was added with the main goal of removing style attributes.

shameondev commented on zerver/lib/send_email.py:
```diff
@@ -137,13 +141,16 @@ def render_templates() -> Tuple[str, str, str]:
-            compiled_template_prefix = os.path.join(emails_dir, "compiled", template)
-            html_message = loader.render_to_string(compiled_template_prefix + ".html", context)
-        return (html_message, message, email_subject)
+            # What should we do here since we don't have compiled htmls anymore?
```
What should we do here since we don't have compiled htmls anymore?

timabbott replied:
So the previous logic was just hackily look for the template in two paths -- the `compiled/` directory and the main one. Since we don't need to compile emails anymore, I suspect you can just get rid of all the complexity -- replace `.source.html` with just `.html` in the filenames, just look at the original path, and not catch the exception -- if the email template doesn't exist, it's a bug.

timabbott commented on zerver/tests/test_email_notifications.py:
```diff
@@ -1338,15 +1338,15 @@ def test_sender_name_in_missed_message(self) -> None:
         # If message content starts with <p> tag the sender name is appended inside the <p> tag.
         self.assertIn(
-            '<p><b>Iago</b>: <span class="user-mention"', mail.outbox[0].alternatives[0][0]
+            '<p style="font-family: sans-serif;font-size: 14px;font-weight: normal;margin: 10px 0;"><b>Iago</b>: <span class="user-mention"', mail.outbox[0].alternatives[0][0]
```
Since it's just an `assertIn`, let's just pull the `p` tag out of the assertion, so we don't have all the CSS in this test.

(This is probably a good option for most of these).

I think we want at a couple tests that include all the interpolated CSS in it in this file, to verify that we are doing the CSS interpolation, but generally it's just going to make a test harder to read and we should just check for `><b>Iago</b>: <span class="user-mention"` to see that this is working as intended? Hmm, that might not work for this test case in particular, because the assertion suggested by the comment IS about the `p` page, but maybe you can apply this idea to the other ones with an image tag or something.

(Or maybe we should just replace this assertion with an HTML assertion using `xpath` or something)

timabbott commented:
The failing tests suggest you get to delete some more code in the `tools/build-release-tarball` code path -- you can run that tool manually in a development environment for a decent debug/refresh cycle for parts of understanding that.  

`git grep /compiled` should be a good thing to read looking for other things to change; I don't think we "compile" much else.

timabbott commented:
> We need to address the usage of inline_email_css in send_custom_email function and come up with a solution for it.

@alexmv can confirm my understanding, but I'm pretty sure that we probably can just delete that logic -- since it uses `build_email/send_email` to send the actual emails, doing the CSS inlining in that step should eliminate the need for the `send_custom_email` code path to directly do anything related to CSS inlining.

alexmv commented:
> > We need to address the usage of inline_email_css in send_custom_email function and come up with a solution for it.
> 
> @alexmv can confirm my understanding, but I'm pretty sure that we probably can just delete that logic -- since it uses `build_email/send_email` to send the actual emails, doing the CSS inlining in that step should eliminate the need for the `send_custom_email` code path to directly do anything related to CSS inlining.

That sounds right to me, yeah.

shameondev commented:
Thank you for the insightful review, folks!

Currently, I have fixed the tests, removed the inlining logic from `send_custom_email`, slightly rewrote the documentation, and fixed the CI. I am pleased with how everything turned out now!

timabbott commented:
This is great, I squashed the commits and wrote a commit message summarizing the benefits of this work (which was helpful at least for me in thinking things through). As a side note, from a review standpoint, there's almost never a benefit to having a separate commit for the test changes for something -- it's always easy to mentally separate those via just looking at different subsets of the files anyway.

Merged, thanks @shameondev!
